File: HW01/22001641_karatas_emre.lua
-- Declare an empty list
local emptyList = {}

-- Initialize a list with some values
local famousCSPeople = {"Tim Barners-Lee", "Alan Turing", "Barbara Liskov"}

-- Check if the list is empty or not
print(#emptyList == 0 and "emptyList is empty" or "emptyList is not empty")
print(#famousCSPeople == 0 and "famousCSPeople is empty" or "famousCSPeople is not empty")

-- Add a new element to a list
table.insert(famousCSPeople, "Lars Bak")

-- Check if a particular element exists in the list
function contains(list, element)
    for _, value in ipairs(list) do
        if value == element then
            return true
        end
    end
    return false
end

if contains(famousCSPeople, "Ada Lovelace") then
    print("Ada Lovelace exists in famousCSPeople")
else
    print("Ada Lovelace does not exist in famousCSPeople")
end

-- Remove a particular element from the list
function removeElement(list, element)
    for i, value in ipairs(list) do
        if value == element then
            table.remove(list, i)
            break
        end
    end
end

removeElement(famousCSPeople, "Barbara Liskov")

-- Get the head and the tail of a list
if #famousCSPeople > 0 then
    local head = famousCSPeople[1]
    local tail = {table.unpack(famousCSPeople, 2)}

    -- Print the head and tail of the list with labels
    print("Head is: " .. head)
    print("Tail is: " .. table.concat(tail, ", "))
else
    print("The famousCSPeople array is empty or not defined.")
end

-- Print all of the elements in the list
if #famousCSPeople > 0 then
    print(table.concat(famousCSPeople, ", "))
else
    print("The famousCSPeople array is empty or not defined.")
end

File: HW01/22001641_karatas_emre.rs
fn main() {
    // Declare/create an empty list
    let mut empty_list: Vec<String> = Vec::new();
    
    // Initialize a list with some values
    let mut famousCSPeople = vec![String::from("Tim Barners-Lee"), String::from("Alan Turing"), String::from("Barbara Liskov")];

    // Check if the list is empty or not
    let is_empty_empty_list = empty_list.is_empty();
    println!("Is the empty list empty? {}", is_empty_empty_list);
    
    // Check if the list is empty or not
    let is_empty_famousCSPeople = famousCSPeople.is_empty();
    println!("Is the famousCSPeople empty? {}", is_empty_famousCSPeople);

    // Add a new element to a list
    famousCSPeople.push(String::from("Lars Bak"));

    // Check if a particular element exists in the list
    let element = "Ada Lovalace";
    let exists = famousCSPeople.contains(&element.to_string());
    println!("Does the famousCSPeople contain {}? {}", element, exists);

    // Remove a particular element from the list
     let element_to_remove = "Barbara Liskov";
    famousCSPeople.retain(|x| x != element_to_remove);

    // Get the head and the tail of a list
    if let Some(head) = famousCSPeople.first() {
        println!("Head of the famousCSPeople: {}", head);
    }
    if famousCSPeople.len() > 1 {
        let tail = &famousCSPeople[1..];
        println!("Tail of the famousCSPeople: {:?}", tail);
    }

    // Print all of the elements in the list
    println!("famousCSPeople elements: {:?}", famousCSPeople);
}

